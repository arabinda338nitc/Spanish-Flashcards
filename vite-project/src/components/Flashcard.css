.flashcard-container {
    perspective: 1000px;
}

.flashcard {
    width: 300px;
    height: 200px;
    position: relative;
    transform-style: preserve-3d;
    transition: transform 0.6s;
    cursor: pointer;
}

.flashcard.flipped {
    transform: rotateY(180deg);
}

.card-face {
    position: absolute;
    width: 100%;
    height: 100%;
    backface-visibility: hidden;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 1.5em;
    border: 1px solid #ccc;
    border-radius: 10px;
    padding: 20px;
    box-sizing: border-box;
}

.card-front {
    background-color: var(--card-front-bg);
}

.card-back {
    background-color: var(--card-back-bg);
    transform: rotateY(180deg);
}

.flashcard-buttons {
    display: flex;
    justify-content: center;
    gap: 16px;
    margin-top: 16px;
}

.flashcard-buttons button {
    padding: 8px 16px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
}

.right-button {
    background-color: var(--right-color);
    color: white;
}

.wrong-button {
    background-color: var(--wrong-color);
    color: white;
}

.helper-text {
    margin-top: 16px;
    color: var(--helper-text-color);
    font-style: italic;
} 