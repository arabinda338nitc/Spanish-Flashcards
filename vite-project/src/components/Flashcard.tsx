import { useState, useEffect } from 'react';
import './Flashcard.css';

interface FlashcardProps {
  spanish: string;
  english: string;
  onRight: () => void;
  onWrong: () => void;
  cardKey?: string;
}

/**
 * Flashcard component
 *
 * This component renders a flashcard with a Spanish word on the front and an English word on the back.
 * The card can be flipped by clicking on it.
 * When the card is flipped, the user can mark their answer as right or wrong.
 */
const Flashcard: React.FC<FlashcardProps> = ({ spanish, english, onRight, onWrong, cardKey }) => {
    // State to track if the card is flipped
    const [flipped, setFlipped] = useState(false);

    // useEffect hook to reset the flipped state when the card changes
    useEffect(() => {
        setFlipped(false);
    }, [cardKey]);

    // CSS classes for the card
    const cardClasses = `flashcard ${flipped ? 'flipped' : ''}`;

    return (
        <div className="flashcard-container">
            <div
                className={cardClasses}
                onClick={() => setFlipped(!flipped)}
                data-testid="flashcard"
            >
                {/* Front of the card */}
                <div className="card-face card-front" data-testid="flashcard-content-front">
                    {spanish}
                </div>
                {/* Back of the card */}
                <div className="card-face card-back" data-testid="flashcard-content-back">
                    {english}
                </div>
            </div>

            {/* Buttons to mark the answer as right or wrong */}
            {flipped && (
                <div className="flashcard-buttons">
                    <button className="right-button" onClick={onRight}>✅ Right</button>
                    <button className="wrong-button" onClick={onWrong}>❌ Wrong</button>
                </div>
            )}

            {/* Helper text to guide the user */}
            {!flipped && (
                <div className="helper-text">
                    Click the card to see the translation
                </div>
            )}
        </div>
    );
};

export default Flashcard; 