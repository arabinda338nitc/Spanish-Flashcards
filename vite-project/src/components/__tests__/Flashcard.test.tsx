import { render, screen, fireEvent } from '@testing-library/react';
import Flashcard from '../Flashcard';

describe('Flashcard', () => {
  const baseProps = {
    spanish: 'el gato',
    english: 'the cat',
    onRight: jest.fn(),
    onWrong: jest.fn(),
    cardKey: 'animals-0',
  };

  it('renders the Spanish word by default', () => {
    render(<Flashcard {...baseProps} />);
    expect(screen.getByText('el gato')).toBeInTheDocument();
  });

  it('flips to show the English translation on click', () => {
    render(<Flashcard {...baseProps} />);
    fireEvent.click(screen.getByTestId('flashcard-content'));
    expect(screen.getByText('the cat')).toBeInTheDocument();
  });

  it('calls onRight when Right button is clicked', () => {
    render(<Flashcard {...baseProps} />);
    fireEvent.click(screen.getByTestId('flashcard-content'));
    fireEvent.click(screen.getByText('✅ Right'));
    expect(baseProps.onRight).toHaveBeenCalled();
  });

  it('calls onWrong when Wrong button is clicked', () => {
    render(<Flashcard {...baseProps} />);
    fireEvent.click(screen.getByTestId('flashcard-content'));
    fireEvent.click(screen.getByText('❌ Wrong'));
    expect(baseProps.onWrong).toHaveBeenCalled();
  });


}); 