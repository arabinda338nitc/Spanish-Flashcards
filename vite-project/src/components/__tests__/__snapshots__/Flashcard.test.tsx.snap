// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Flashcard matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column; align-items: center;"
  >
    <div
      style="width: 300px; height: 180px; background: rgb(255, 255, 255); border-radius: 12px; box-shadow: 0 2px 8px rgba(0,0,0,0.1); display: flex; align-items: center; justify-content: center; font-size: 2em; cursor: pointer; margin-bottom: 24px; transition: all 0.3s ease; transform: scale(1); border: 2px solid transparent;"
    >
      <div
        style="text-align: center; transition: all 0.3s ease;"
      >
        el gato
      </div>
    </div>
    <div
      style="margin-top: 16px; color: rgb(102, 102, 102); font-size: 14px; text-align: center;"
    >
      Click the card to see the translation
    </div>
  </div>
</DocumentFragment>
`;
